common/src/main/java/io/pravega/common/util/BufferViewOutputStream.java

```diff
@@ -0,0 +1,280 @@
+/**
+ * Copyright (c) Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.common.util;
+
+import com.google.common.annotations.VisibleForTesting;
+import io.pravega.common.Exceptions;
+import java.io.DataOutput;
+import java.io.OutputStream;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+import javax.annotation.concurrent.NotThreadSafe;
+
+@NotThreadSafe
+public final class BufferViewOutputStream extends OutputStream implements DataOutput {
```

Comment:
I thought about that, but I think the uses are sufficiently different. `BufferViewOutputStream` has no requirement to read and is just building output. So it can be efficient by avoiding copies when an array is appended. If I were to add the functionality to `CompositeByteArraySegment` it would mean that all of the random access read methods such as `get` would no longer be `O(1)` and I didn't consider that a worthwhile tradeoff.